src/main/java/io/netty/incubator/codec/http3/QpackDecoder.java

```diff
@@ -0,0 +1,223 @@
+        // https://tools.ietf.org/html/draft-ietf-quic-qpack-19#section-4.5.1.2
+        decodePrefixedInteger(in, 7);
+
+        while (in.isReadable()) {
+            byte b = in.getByte(in.readerIndex());
+            if ((b & 0x80) == 0x80) {
+                // 1xxxxxxx
+                decodeIndexed(in, sink);
+            } else if ((b & 0xc0) == 0x40) {
+                // 01xxxxxx
+                decodeLiteralWithNameReference(in, sink);
+            } else if ((b & 0xe0) == 0x20) {
+                // 001xxxxx
+                decodeLiteral(in, sink);
+            }
+        }
+    }
+
+    private long decodePrefixedInteger(ByteBuf in, int prefixLength) {
+        int nbits = (1 << prefixLength) - 1;
+        int first = in.readByte() & nbits;
+        if (first < nbits) {
+            return first;
+        }
+
+        long i = first;
+        int factor = 0;
+        byte next;
+        do {
+            next = in.readByte();
+            i += (next & 0x7f) << factor;
+            factor += 7;
+        } while ((next & 0x80) == 0x80);
+
+        return i;
+    }
+
+    private void decodeIndexed(ByteBuf in, Sink sink) throws Http3Exception {
+        if ((in.getByte(in.readerIndex()) & 0x40) == 0x40) {
+            final int staticIndex = (int) decodePrefixedInteger(in, 6);
+            final QpackHeaderField field = getIndexedHeader(staticIndex);
+            sink.appendToHeaderList(field.name, field.value);
+        } else {
+            throw new UnsupportedOperationException("dynamic table is not implemented");
+        }
+    }
+
+    private void decodeLiteralWithNameReference(ByteBuf in, Sink sink) throws Http3Exception {
+        if ((in.getByte(in.readerIndex()) & 0x10) == 0x10) {
+            final int staticNameIndex = (int) decodePrefixedInteger(in, 4);
+            final CharSequence name = getIndexedName(staticNameIndex);
+            final CharSequence value = decodePrefixedStringLiteral(in);
+            sink.appendToHeaderList(name, value);
+        } else {
+            throw new UnsupportedOperationException("dynamic table is not implemented");
+        }
+    }
+
+    private void decodeLiteral(ByteBuf in, Sink sink) throws Http3Exception {
+        final CharSequence name = decodePrefixedStringLiteral(in, (byte) 0x8, 3);
+        final CharSequence value = decodePrefixedStringLiteral(in);
+        sink.appendToHeaderList(name, value);
+    }
+
+    private CharSequence decodePrefixedStringLiteral(ByteBuf in) throws Http3Exception {
+        return decodePrefixedStringLiteral(in, (byte) 0x80, 7);
+    }
+
+    private CharSequence decodePrefixedStringLiteral(ByteBuf in, byte mask, int prefix) throws Http3Exception {
+        final boolean huffmanEncoded = (in.getByte(in.readerIndex()) & mask) == mask;
+        final int length = (int) decodePrefixedInteger(in, prefix);
+        return decodeStringLiteral(in, length, huffmanEncoded);
+    }
+
+    // TODO: not sure if integer length is enough
```

Comment:
can you explain what the problem is ?

Reply:
so... regarding parsing prefixed integers, the spec mentions [here](https://tools.ietf.org/html/draft-ietf-quic-qpack-19#section-4.1.1)

>    QPACK implementations MUST be able to decode integers up to and including 62 bits long.

That is why I've  made `decodePrefixedInteger` to return `long`. But when I pass decoded length as the parameter for decoding string literal, I convert it to `int`. Simply because I do `new byte[length]` in there. Though  I don't think it's realistic to get more than int bytes in one message, I didn't find anything in spec that prevents  peer from doing so. It would be safe to remove TODO as of now, it was mostly meant for me to do some digging

Reply:
Yeah I think its fine...